sklearn/utils/tests/test_validation.py

```diff
@@ -975,23 +977,54 @@ def test_check_is_fitted_with_attributes(wrap):
         check_is_fitted(ard, wrap(["coef_bad_"]))
 
 
-def test_check_consistent_length():
-    check_consistent_length([1], [2], [3], [4], [5])
-    check_consistent_length([[1, 2], [[1, 2]]], [1, 2], ["a", "b"])
+def test_check_consistent_length_sparse():
+    """Test that `check_consistent_length` works with sparse data."""
     check_consistent_length([1], (2,), np.array([3]), sp.csr_matrix((1, 2)))
+
+
+@pytest.mark.parametrize(
+    "array_namespace, device, _", yield_namespace_device_dtype_combinations()
+)
+def test_check_consistent_length(array_namespace, device, _):
+    """Test that `check_consistent_length` raises on inconsistent lengthes and wrong
+    input types trigger TypeErrors."""
+    xp = _array_api_for_tests(array_namespace, device)
+
+    check_consistent_length(
+        xp.asarray([1], device=device),
+        xp.asarray([2], device=device),
+        xp.asarray([3], device=device),
+        xp.asarray([4], device=device),
+        xp.asarray([5], device=device),
+    )
+    if xp.__name__ == "numpy":
+        check_consistent_length(
+            xp.asarray([[1, 2], [1, 2]], device=device),
+            xp.asarray([1, 2], device=device),
+            xp.asarray(["a", "b"], device=device),
+        )
+    else:
+        check_consistent_length(
+            xp.asarray([[1, 2], [1, 2]], device=device),
+            xp.asarray([1, 2], device=device),
+        )
     with pytest.raises(ValueError, match="inconsistent numbers of samples"):
-        check_consistent_length([1, 2], [1])
+        check_consistent_length(
+            xp.asarray([1, 2], device=device), xp.asarray([1], device=device)
+        )
     with pytest.raises(TypeError, match=r"got <\w+ 'int'>"):
-        check_consistent_length([1, 2], 1)
+        check_consistent_length(xp.asarray([1, 2], device=device), 1)
     with pytest.raises(TypeError, match=r"got <\w+ 'object'>"):
-        check_consistent_length([1, 2], object())
+        check_consistent_length(xp.asarray([1, 2], device=device), object())
 
     with pytest.raises(TypeError):
-        check_consistent_length([1, 2], np.array(1))
+        check_consistent_length(xp.asarray([1, 2], device=device), np.array(1))
```

Comment:
this wouldn't raise when `xp==np`, would it?

Reply:
I have just tried it. I raises with `TypeError: Singleton array array(1) cannot be considered a valid collection.` from within `_num_samples() `in sklearn/utils/validation.py:376.

The last four pytest.raises checks are not really testing `check_consistent_length()`, but some other error mechanism outside of it. It feels not very clean. What would be a good way to handle this? Run those tests separately?